### ISSUE TITLE
Analyze Phenotype: workflow redesign

### ISSUE THREAD

**Issue by yuanzhou** [monarch-initiative/monarch-legacy#1250]
Based on @harryhoch's mockups, I started the redesign. The first thing I did was to clean up the page and related code. I also rephrased the section titles as you can see from the second image. In section of step 2, instead of organizing the search and compare side by side, I made them vertically aligned, so the UI looks cleaner. I've also removed the reset button since it doesn't work well. And made the submit button lots bigger. 

**OLD version:**

![old](https://cloud.githubusercontent.com/assets/195873/15399428/913d4c70-1db5-11e6-961e-72efb8582373.PNG)

**NEW version:**

![new](https://cloud.githubusercontent.com/assets/195873/15399432/93823450-1db5-11e6-8f81-1cd2c18c8b03.PNG)

@harryhoch @jmcmurry please add comments and I'll keep refining. I created this issue to encourage discussions during this redesign. This is just the starting point.

**Comment by yuanzhou** [comment]
Made some wording changes. Are "Select Comparables" and "Selected Comparables" appropriate?

![1](https://cloud.githubusercontent.com/assets/195873/15547268/13433b98-2271-11e6-9210-42e85d1c995e.PNG)

And added the selected inputs on the result page:

![2](https://cloud.githubusercontent.com/assets/195873/15547269/15b7206a-2271-11e6-93c4-eae79a75b76b.PNG)

**Comment by harryhoch** [comment]
fine. Are you going to make the selected phenotypes removable?

**Comment by yuanzhou** [comment]
@harryhoch I'm going to play with that. Every time we remove one phenotype, the whole page will need to be reloaded to refresh Phenogrid, which also means we'll have to wait for the owlsim querying. Not sure if there's a more efficient way of doing this.

**Comment by harryhoch** [comment]
@yuanzhou, not necessarily. Perhaps include a "refresh" button which would only be enabled once someone changes the phenotypes. No update of phenogrid until that button is pressed...

**Comment by yuanzhou** [comment]
@harryhoch One thing I realized is the selected phenotype IDs are in the URL, for example: 

```
http://localhost:8484/analyze/phenotypes/?input_items=HP%3A0001937+HP%3A0012594+MP%3A0009235&mode=search&target_species=all&limit=100
```

There are 3 phenotypes, and when I make the phenotypes removable, we'll also need to update the URL to refresh the result, including Phenogrid.

**Comment by harryhoch** [comment]
would it help to remove the phenotypes from the URL and to go to a PUT instead of a GET? Might require some monkeying on the middleware, but shouldn't be too bad.

**Comment by jmcmurry** [comment]
My feeling is that both the original and modified versions of this interface plunge the user in a bit too fast into 'advanced' mode. I would love to see less emphasis on the homologs/species and more on dynamic results and faceting as previously discussed.

![analyzephenotypes_redesign_general_2016-06-03-01](https://cloud.githubusercontent.com/assets/3658863/15792295/60a12388-29aa-11e6-962a-65327cfdce23.png)

**Comment by harryhoch** [comment]
I don't disagree. Should we rethink efforts/priorities?

**Comment by kshefchek** [comment]
In response to https://github.com/monarch-initiative/monarch-app/issues/1250#issuecomment-223682134, I think it makes sense to have this page be completely client driven, the server just provides the table view, and that can easily be moved to the browser - and probably needs some work anyway.

We shouldn't remove the URL params entirely, because this is still used from the annotate text page to send a list of phenotypes to the analyze page.  But once on the page it should be single page app-like.

**Comment by harryhoch** [comment]
@kshefchek , it can work both ways, right? we can enter the page with a get and then change content from there based on user interaction...

**Comment by kshefchek** [comment]
Yep, we're on the same page!

**Comment by jmcmurry** [comment]
@yuanzhou what are all the contents of this dropdown menu currently?
<img width="186" alt="screen shot 2016-06-07 at 2 42 03 pm" src="https://cloud.githubusercontent.com/assets/3658863/15875674/171a0894-2cbe-11e6-8334-417c3b9033a8.png">

**Comment by jmcmurry** [comment]
It is my opinion that the two most crucial things to fix are:
1) Combine inputs / results on one (dynamic) page
2) Make genes, models, diseases discrete sets of results (each of these solves a different use case). Conflating them is more confusing than needs to be.

Both of these objectives are steps the path to the more comprehensive redesign originally proposed.

I've taken a stab at how we might accomplish the above two objectives with the existing widgets, but would like to know whether objective 2 is problematic from the backend standpoint and phenogrid more generally. I would like if possible not to have a refresh button for phenogrid, but we can certainly revisit if it is too intense. The table parts should be easily made dynamic. Thoughts?

![analyzephenotypes_redesign_general_2016-06-07b-01](https://cloud.githubusercontent.com/assets/3658863/15941461/1ff24698-2e35-11e6-8c50-aa4308fb211a.png)

How do people feel about this as a first pass? Is it doable before July? @harryhoch @cmungall

**Comment by harryhoch** [comment]
Julie:
A few comments.
1. Goals one and two sound reasonable. 
2. Are you suggesting populating the discrete models, genes, and diseases based on search terms? Seems reasonable to me, but I must defer to @cmungall , @yuanzhou , @cborromeo , and @kshefchek  for feasibility. 
3. Not sure what you mean by "not having a refresh button for phenogrid." At the moment, I think we can't have an automatic refresh. Simsearch calls take a few seconds to return values, so we don't want the delay triggering unless users explicitly request.

As far as doable by July, I'm skeptical. Possible, but I don't like rushing these things. I'm also skeptical that this can be divvied up amongst multiple folks. @yuanzhou, what do you think?

**Comment by yuanzhou** [comment]
@jmcmurry in my branch, the current drop down contains 
- All
- Mus musculus (genes)
- Danio rerio (genes)
- Homo sapiens (diseases)
- Drosophila melanogaster (genes)

We disabled the "Caenorhabditis elegans (genes)"

**Comment by yuanzhou** [comment]
@harryhoch I agree with you regarding the automatic Phenogrid refresh. Sometimes simsearch calls take more than several seconds.

I'm not positive that I can get this done by July considering I'm also tied to several other projects. I'll try my best. And a clear design goal will definitely be very helpful to avoid miscommunication and guess work.

**Comment by yuanzhou** [comment]
@jmcmurry regarding your second point:

> 2) Make genes, models, diseases discrete sets of results (each of these solves a different use case). Conflating them is more confusing than needs to be.

Are you suggesting to have a rendered Phenorgrid under the data table at each discrete section?

As @harryhoch mentioned, I'll probably need some input from @kshefchek

**Comment by jmcmurry** [comment]
> Are you suggesting to have a rendered Phenorgrid under the data table at each discrete section?

Yes, that's my proposal but not all three phenogrids in one page at one time. Just one phenogrid, corresponding to whichever tab the user clicks (Genes, Diseases, Models). A side-benefit of this decoupled approach is that there would be fewer species to render. (Diseases would be human only, models would be mostly not human)

**Comment by cmungall** [comment]
> Sometimes simsearch calls take more than several seconds

We need to reduce this. I suspect the owlsim call itself is fast, but the app wrapper is doing other stuff

**Comment by jmcmurry** [comment]
@cmungall what is your preference? Should we:
A) Aim to improve response times before July OR
B) Focus on UI improvements and do a "refresh" button as a short-term fix

Without dynamic results, I'm not convinced there's much to be gained by having inputs/outputs on the same tab. Regardless of where we go with the interface, performance would need to be improved anyway. By reducing the scope of any one call to a particular type of result (Gene, Disease, or Model), wouldn't performance be better anyway (since you're reducing the search space)?

**Comment by cmungall** [comment]
There is no model search in the current owlsim2 instance. Only human d2p and model organism g2p associations are loaded.

we need more data before making a final decision. Can we simulate the calls required in a typical session in a script and do some benchmarks using that?

**Comment by harryhoch** [comment]
@cmungall, I might be able to do something quick and dirty tonight. will see what I can come up with.

**Comment by harryhoch** [comment]
@cmungall, just did some rough checks. Saw response times to POSTS to https://monarchinitiative.org/simsearch/phenotype ranging from .001s to .925 sec. We'd like to see < 200ms for perception of instantaneous update.

**Comment by jmcmurry** [comment]
Related to https://github.com/monarch-initiative/monarch-app/issues/1238; please check and close if appropriate. Thanks

**Comment by harryhoch** [comment]
ok. tried further running against local monarch-app with no cache on results .  I'm seeing calls to https://monarchinitiative.org/owlsim/searchByAttributeSet going from .1 sec to 3.2 sec. 
@cmungall, these are the calls that go straight to OWLSim, right?

**Comment by jmcmurry** [comment]
So having spoken, we decided we would aim to 
1) Put inputs, phenogrid, and table on one (visual) page
2) Provide a refresh button for phenogrid and table
Ultimately, it would be good to get away from this, but whatever we do, let's not do [as MBTA does](http://www.mbta.com/schedules_and_maps/rail/lines/?route=PROVSTOU): They not only require a click to update but they continue to display data brazenly without even indicating it no longer corresponds to the (revised) user inputs. This too is strange: their REST API has "RedisplayTime=Redisplay+Time" in it. As if stale data would be worth having.

<img width="566" alt="screen shot 2016-06-09 at 11 48 51 am" src="https://cloud.githubusercontent.com/assets/3658863/15942507/8f431e74-2e39-11e6-8973-4bf18acc9a01.png">

We could either:
A) Have some kind of greying out of the table and grid to flag the fact that it is necessary to click to update.
B) Change the appearance of the input UI so that it clearly has two modes, active and inactive. An example of a site that I think does this well is Kayak.
<img width="801" alt="screen shot 2016-06-09 at 12 11 52 pm" src="https://cloud.githubusercontent.com/assets/3658863/15942962/9ed6b6b4-2e3b-11e6-8344-fef0fad95e99.png">
<img width="812" alt="screen shot 2016-06-09 at 12 11 42 pm" src="https://cloud.githubusercontent.com/assets/3658863/15942964/a0aa3f74-2e3b-11e6-981f-a108023b7c62.png">

Thoughts?

**Comment by harryhoch** [comment]
also, @jmcmurry suggests hiding the advanced options in a accordion style display so they aren't seen unless needed.

**Comment by yuanzhou** [comment]
Any strong opinion that we should keep the "Orthologs" and "Paralogs" buttons? I made progress with the one page design, I'm also trying to make the comparables section as short as possible. And since we are having that "discrete sets of result" idea in mind, it might be helpful to focus fewer input aspects as a new starting point.

**Comment by kshefchek** [comment]
It's an important use case, we need to either leave it in this widget or make the option available elsewhere.

**Comment by jmcmurry** [comment]
Agreed we need to keep them. Perhaps hiding within an "advanced" button
though?

**Comment by harryhoch** [comment]
+1

**Comment by yuanzhou** [comment]
So far I've got the one page working.

![1](https://cloud.githubusercontent.com/assets/195873/16086037/23f3c8b4-32ec-11e6-951f-e462379fcf2e.PNG)

The orthologos and paralogs buttons will show up only when the genes option is selected.
![capture](https://cloud.githubusercontent.com/assets/195873/16086086/4d83b018-32ec-11e6-9758-608d3cb4693e.PNG)

Once the submit button is clicked, the phenogrid and data table will show below the button. Whenever there's a new phenotype is added or one gets removed from the selected list, the phenogrid and data bale will fade out. This encourages the users to reclick the submit button to get new results.

**Comment by jmcmurry** [comment]
> The orthologos and paralogs buttons will show up only when the genes option is selected.

Thanks Zhou. This is a great start; I also like the following you added:
- The different language of "Predefined target groups."
- That step 1 is the full width of the screen (This emphasizes the right things first and will make the list of phenotypes less vertically space hungry).

I have a few suggestions in rough order of importance:
- I would suggest making the buttons themselves smaller as it will make it more obvious that they're buttons.
- Make the text on the buttons just say "submit" and in larger font
- Change the first section text to "create a profile of phenotypes"
- In the predefined target groups, I'm not sure what the default should say. I get that having a default called "all" is a bit weird to the uninitiated user; Could we make the default "all genes, all diseases" (thoughts on this? I'm not wild about making mouse genes the default as this sells the platform short.)
- I would like if we could arrange the elements to make better use of vertical space. This way we don't need to do extra to direct the users attention to "below the fold". Suggestion below, side by side to illustrate the issue. The red line indicates where the 'fold' lands.

<img width="763" alt="screen shot 2016-06-15 at 1 38 03 pm" src="https://cloud.githubusercontent.com/assets/3658863/16096560/74fd605a-32fe-11e6-9270-71a54c65007f.png">
- I would like it if we could have a canned query people could execute by clicking "example query" (I didn't mock this up but wanted to know what others thought).
- The behavior I'd suggest when someone fiddles with inputs after phenogrid has been rendered would be to 1) decrease opacity of phenogrid and table, 2) change text on the button to "refresh" 3) change color on button to orange in order to make sure it is noticed. (Shown below)
  <img width="874" alt="screen shot 2016-06-15 at 1 56 53 pm" src="https://cloud.githubusercontent.com/assets/3658863/16097142/0c721640-3301-11e6-9343-07c905b7377c.png">

**Comment by harryhoch** [comment]
@yuanzhou, I agree - this is a great start.

@jmcmurry, I like your changes as well. Zhou, do you think you can add these changes in? Thanks.

**Comment by yuanzhou** [comment]
@jmcmurry Thanks for the great suggestions, one thing to clarify:

> Make the text on the buttons just say "submit" and in larger font

It says "Analyze" in your mockup, which one do you prefer?

@harryhoch I'm working on those changes. And regarding the Step 2, I also think it might be better to display those species in checkboxes instead of dropdown menu. Checkbox allows users to select any combinations they want to analyze against. Right now the dropdown only gives options of single species or all. In some use cases, e.g., user may want only see mouse or human. Even though you can enable/disable species via the "Options" button of Phenogrid, but this won't update the data table.

**Comment by harryhoch** [comment]
@yuanzhou Ok.  If space is at a premium, you can put checkboxes in a dropown.

**Comment by jmcmurry** [comment]
Thanks
1) As for submit vs analyze I don't feel strongly. Harry? Others?
2) As for checkboxes, that is a good idea. We can probably spare a little vertical space for a few rows. Expand-collapse dialog with checkboxes could be fine too. Hard to know until we look at it.

**Comment by harryhoch** [comment]
I'll go for "analyze", but not for any strong reason

**Comment by cmungall** [comment]
> 1) As for submit vs analyze I don't feel strongly. Harry? Others?

Search?

**Comment by mellybelly** [comment]
Well the title of the page is "Monarch Phenotype Profile Analysis", so maybe "Analyze"?

**Comment by jmcmurry** [comment]
Sounds good. I don't love 'search' as that sounds a bit pedestrian. OwlSim is anything but :) 
Go, OwlSim, Go!

**Comment by yuanzhou** [comment]
After the demo this afternoon, @jmcmurry suggested to add back the gene helper text as well as link the gene ID section to https://github.com/monarch-initiative/dipper#identifiers.

![capture2](https://cloud.githubusercontent.com/assets/195873/16318812/b4baf1c4-395e-11e6-8b44-1c3f5b26f635.PNG)

I've also made the results grayed out on adding/removing new phenotypes, and changed the submit button color and text to highlight the change.

![capture](https://cloud.githubusercontent.com/assets/195873/16318754/859e291a-395e-11e6-9402-c358f124ac4e.PNG)

**Comment by harryhoch** [comment]
Nice, @yuanzhou , thanks
